lib/resty/wasmx/proxy_wasm.lua

```diff
@@ -38,23 +38,32 @@ local _M = {
 
 if subsystem == "http" then
     ffi.cdef [[
-        int ngx_http_wasm_ffi_plan_new(ngx_wasm_vm_t *vm,
-                                       ngx_wasm_filter_t *filters,
-                                       size_t n_filters,
-                                       ngx_wasm_plan_t **out,
-                                       u_char *err, size_t *errlen);
-        int ngx_http_wasm_ffi_plan_free(ngx_wasm_plan_t *plan);
-        int ngx_http_wasm_ffi_plan_load(ngx_wasm_plan_t *plan);
-        int ngx_http_wasm_ffi_plan_attach(ngx_http_request_t *r,
-                                          ngx_wasm_plan_t *plan,
-                                          unsigned int isolation);
-        int ngx_http_wasm_ffi_start(ngx_http_request_t *r);
-        int ngx_http_wasm_ffi_set_property(ngx_http_request_t *r,
-                                           ngx_str_t *key,
-                                           ngx_str_t *value);
-        int ngx_http_wasm_ffi_get_property(ngx_http_request_t *r,
-                                           ngx_str_t *key,
-                                           ngx_str_t *out);
+        ngx_int_t ngx_http_wasm_ffi_plan_new(ngx_wasm_vm_t *vm,
```

Comment:
There is: `sizeof(int) == 4`, `sizeof(ngx_int_t) == 8`. Using `int` in a function signature that required `ngx_int_t` caused me signal extension errors in the C side when propagating negative error codes. So since I needed to use the correct type for the function I was hitting, I figured I'd fix it as well for all the other ones, lest a similar bug might be lurking somewhere.

Reply:
Right but... int being signed, which return value exactly? 🤔

Reply:
It happened to `rc` values other than `NGX_OK`, which are negative, such as `NGX_DECLINED` (-5). IIRC it happened when passing that value through the LuaJIT callback typedef. That signed negative 32-bit number would auto-expand into a 64-bit value without sign-extension happening, so I was seeing a high positive number (typical 4-billion-something, probably 2^32-5) in the C side where -5 was expected, which fixed itself when I changed the FFI definition on the Lua side from `int` to the correct definition of `ngx_int_t` (`intptr_t`, which is 64-bit and predefined by LuaJIT).

Reply:
I do not understand how that is possible. We already have test cases with negative return values such as `NGX_ERROR` (-1) and `NGX_ABORT` (-6), and `int` should already be signed and able to carry these error values. Something else is probably at play here?

Reply:
> Something else is probably at play here?

I believe it's LuaJIT's autoconversion of Lua closure into C function pointers. This is new.

Reply:
If we don't need this anymore I'd be in favor of reverting it to keep the FFI as simple as possible.